File: src/aggregate.ts
/**
 * 聚合模型：总消耗量窗口指标。
 * 统计口径（spec 决策锚）：仅 type=message && role=assistant 且携带 usage 的消息；
 * 总 token 按组件和计算；花费直接累加 cost.total；缓存率先求和分子分母再除。
 */

/** 一条计入口径的 usage（assistant 消息的 message.usage） */
export interface Usage {
  input?: unknown;
  output?: unknown;
  cacheRead?: unknown;
  cacheWrite?: unknown;
  reasoning?: unknown;
  totalTokens?: unknown;
  cost?: { total?: unknown };
}

export interface Totals {
  /** 请求数 = 带 usage 的 assistant 消息数（含全 0 失败消息） */
  requests: number;
  input: number;
  output: number;
  cacheRead: number;
  cacheWrite: number;
  reasoning: number;
  /** 总 token = input + output + cacheRead + cacheWrite（组件和） */
  totalTokens: number;
  /** 花费 = Σ usage.cost.total */
  cost: number;
  /** 缓存率 = cacheRead / (input + cacheRead + cacheWrite)，分母 0 记 0 */
  cacheRate: number;
}

export function emptyTotals(): Totals {
  return {
    requests: 0,
    input: 0,
    output: 0,
    cacheRead: 0,
    cacheWrite: 0,
    reasoning: 0,
    totalTokens: 0,
    cost: 0,
    cacheRate: 0,
  };
}

/** 累加一条计入口径的 usage（assistant 消息的 message.usage） */
export function addUsage(totals: Totals, usage: Usage): void {
  totals.requests += 1;
  totals.input += toFiniteNumber(usage.input);
  totals.output += toFiniteNumber(usage.output);
  totals.cacheRead += toFiniteNumber(usage.cacheRead);
  totals.cacheWrite += toFiniteNumber(usage.cacheWrite);
  totals.reasoning += toFiniteNumber(usage.reasoning);
  totals.cost += toFiniteNumber(usage.cost?.total);
}

/** 聚合收尾：总 token 按组件和、缓存率按分子和/分母和 */
export function finalizeTotals(totals: Totals): void {
  totals.totalTokens = totals.input + totals.output + totals.cacheRead + totals.cacheWrite;
  const denominator = totals.input + totals.cacheRead + totals.cacheWrite;
  totals.cacheRate = denominator === 0 ? 0 : totals.cacheRead / denominator;
}

/** 非有限数字一律按 0 处理（字段缺失 / undefined / null / 非数值） */
function toFiniteNumber(v: unknown): number {
  return typeof v === "number" && Number.isFinite(v) ? v : 0;
}

File: internal/domain/aggregate.go
package domain

import "math"

// Usage 表示一条计入口径的 assistant 消息消耗 (message.usage)
type Usage struct {
	Input       float64 `json:"input"`
	Output      float64 `json:"output"`
	CacheRead   float64 `json:"cacheRead"`
	CacheWrite  float64 `json:"cacheWrite"`
	Reasoning   float64 `json:"reasoning"`
	TotalTokens float64 `json:"totalTokens"`
	Cost        *Cost   `json:"cost,omitempty"`
}

type Cost struct {
	Total float64 `json:"total"`
}

// Totals 表示聚合汇总指标
type Totals struct {
	Requests    int     `json:"requests"`
	Input       float64 `json:"input"`
	Output      float64 `json:"output"`
	CacheRead   float64 `json:"cacheRead"`
	CacheWrite  float64 `json:"cacheWrite"`
	Reasoning   float64 `json:"reasoning"`
	TotalTokens float64 `json:"totalTokens"`
	Cost        float64 `json:"cost"`
	CacheRate   float64 `json:"cacheRate"`
	CostStatus  string  `json:"costStatus,omitempty"`
}

type GroupBy string

const (
	GroupByModel    GroupBy = "model"
	GroupByCwd      GroupBy = "cwd"
	GroupByModelCwd GroupBy = "model,cwd"
)

type Period string

const (
	PeriodDay   Period = "day"
	PeriodWeek  Period = "week"
	PeriodMonth Period = "month"
)

type SessionRow struct {
	Totals
	SessionId       string `json:"sessionId"`
	Timestamp       string `json:"timestamp"`
	Cwd             string `json:"cwd"`
	Model           string `json:"model"`
	FileName        string `json:"fileName,omitempty"`
	DisplayName     string `json:"displayName,omitempty"`
	CwdNorm         string `json:"cwdNorm,omitempty"`
	IsTask          bool   `json:"isTask,omitempty"`
	ParentSessionId string `json:"parentSessionId,omitempty"`
	Source          string `json:"source,omitempty"`
}

type RequestRow struct {
	Totals
	SessionId       string `json:"sessionId"`
	Timestamp       string `json:"timestamp"`
	Model           string `json:"model"`
	DisplayName     string `json:"displayName,omitempty"`
	Source          string `json:"source,omitempty"`
	SourceSessionId string `json:"sourceSessionId,omitempty"`
}

type GroupRow struct {
	Totals
	Model string `json:"model,omitempty"`
	Cwd   string `json:"cwd,omitempty"`
}

type PeriodRow struct {
	Totals
	Period string `json:"period"`
}

func EmptyTotals() Totals {
	return Totals{}
}

func AddUsage(t *Totals, u Usage) {
	t.Requests++
	t.Input += toFinite(u.Input)
	t.Output += toFinite(u.Output)
	t.CacheRead += toFinite(u.CacheRead)
	t.CacheWrite += toFinite(u.CacheWrite)
	t.Reasoning += toFinite(u.Reasoning)
	if u.Cost != nil {
		t.Cost += toFinite(u.Cost.Total)
	}
}

// FinalizeTotals 收尾计算：根据 ADR-0002，总 token = input + cacheRead + output；缓存率 = cacheRead / (input + cacheRead)
func FinalizeTotals(t *Totals) {
	totalInput := t.Input + t.CacheRead
	t.TotalTokens = totalInput + t.Output
	if totalInput == 0 {
		t.CacheRate = 0
	} else {
		t.CacheRate = t.CacheRead / totalInput
	}
}

func toFinite(v float64) float64 {
	if math.IsNaN(v) || math.IsInf(v, 0) {
		return 0
	}
	return v
}
